entities/storobj/storage_object.go

```diff
@@ -432,26 +434,80 @@ func (ko *Object) MarshalBinary() ([]byte, error) {
 	if data[0] != uint8(1) {
 		return errors.Errorf("unsupported binary marshaller version %d", data[0])
 	}
 
+	// clear out old values in case an object misses values. This should NOT shrink the capacity of the map, eg there
+	// are no allocations when adding the properties of the next object again
+	for k := range *properties {
+		delete(*properties, k)
+	}
+
 	startPos := uint64(1 + 8 + 1 + 16 + 8 + 8) // elements at the start
 	byteOps := byte_operations.ByteOperations{Position: startPos, Buffer: data}
 	// get the length of the vector, each element is a float32 (4 bytes)
 	vectorLength := uint64(byteOps.ReadUint16())
 	byteOps.MoveBufferPositionForward(vectorLength * 4)
 
-	// length of class name
 	classnameLength := uint64(byteOps.ReadUint16())
 	byteOps.MoveBufferPositionForward(classnameLength)
-
-	// property schema length
 	propertyLength := uint64(byteOps.ReadUint32())
-	if err := json.Unmarshal(data[byteOps.Position:byteOps.Position+propertyLength], properties); err != nil {
-		return err
-	}
+
+	jsonparser.EachKey(data[byteOps.Position:byteOps.Position+propertyLength], func(idx int, value []byte, dataType jsonparser.ValueType, err error) {
+		switch dataType {
+		case jsonparser.Number:
+			v, _ := jsonparser.ParseFloat(value)
+			(*properties)[aggregationProperties[idx]] = v
+		case jsonparser.Boolean:
+			b, _ := jsonparser.ParseBoolean(value)
+			(*properties)[aggregationProperties[idx]] = b
+		case jsonparser.String:
+			s, _ := jsonparser.ParseString(value)
+			(*properties)[aggregationProperties[idx]] = s
+		case jsonparser.Array: // can be a beacon or an actual array
+			ArrayEntries := value[1 : len(value)-1] // without leading and trailing []
+			beaconVal, errBeacon := jsonparser.GetUnsafeString(ArrayEntries, "beacon")
+			if errBeacon == nil {
+				(*properties)[aggregationProperties[idx]] = []interface{}{map[string]interface{}{"beacon": beaconVal}}
+			} else {
+				// check how many entries there are in the array by counting the ",". This allows us to allocate an
+				// array with the right size without extending it with every append.
+				// The size can be too large for string arrays, when they contain "," as part of their content.
+				entryCount := 0
+				for _, b := range ArrayEntries {
+					if b == uint8(44) { // ',' as byte
+						entryCount++
+					}
+				}
+
+				array := make([]interface{}, 0, entryCount)
+				jsonparser.ArrayEach(value, func(innerValue []byte, innerDataType jsonparser.ValueType, offset int, innerErr error) {
+					var val interface{}
+					var errParse error
+					switch innerDataType {
+					case jsonparser.Number:
+						val, errParse = jsonparser.ParseFloat(innerValue)
+					case jsonparser.String:
+						val, errParse = jsonparser.ParseString(innerValue)
+					case jsonparser.Boolean:
+						val, errParse = jsonparser.ParseBoolean(innerValue)
+					default:
+						panic("Unknown data type ArrayEach") // returning an error would be better
+					}
```

Comment:
maybe this part could be extracted to a separate method (with usage of generics)? and used here and a little bit above where you are also parsing those `Number` `String` and `Boolean` types?

Reply:
How would you use generics here? Depending on the type I need to call different functions from jsonparser?

I think a seperate method will not work, because I have 4 cases above (the 3 you mentioned + array) and only 3 here?

Reply:
I did a second look at this and I might be wrong about using generics for this here... hmm, maybe let's merge it how it is now, and maybe make a follow up task in our next sprint to revisit this implementation, wdyt?